go/test/endtoend/vtgate/queries/reference/main_test.go

```diff
@@ -0,0 +1,292 @@
+			PRIMARY KEY(id)
+		) ENGINE=InnoDB;
+	`
+	shardedVSchema = `
+		{
+			"sharded": true,
+			"vindexes": {
+				"hash": {
+					"type": "hash"
+				}
+			},
+			"tables": {
+				"delivery_failure": {
+					"columnVindexes": [
+						{
+							"column": "id",
+							"name": "hash"
+						}
+					]
+				},
+				"zip_detail": {
+					"type": "reference",
+					"source": "` + unshardedKeyspaceName + `.zip_detail"
+				}
+			}
+		}
+	`
+)
+
+func TestMain(m *testing.M) {
+	defer cluster.PanicHandler(nil)
+	flag.Parse()
+
+	exitCode := func() int {
+		clusterInstance = cluster.NewCluster(cell, hostname)
+		defer clusterInstance.Teardown()
+
+		// Start topo server
+		if err := clusterInstance.StartTopo(); err != nil {
+			return 1
+		}
+
+		// Start keyspace
+		uKeyspace := &cluster.Keyspace{
+			Name:      unshardedKeyspaceName,
+			SchemaSQL: unshardedSQLSchema,
+			VSchema:   unshardedVSchema,
+		}
+		if err := clusterInstance.StartUnshardedKeyspace(*uKeyspace, 0, false); err != nil {
+			return 1
+		}
+
+		sKeyspace := &cluster.Keyspace{
+			Name:      shardedKeyspaceName,
+			SchemaSQL: shardedSQLSchema,
+			VSchema:   shardedVSchema,
+		}
+		if err := clusterInstance.StartKeyspace(*sKeyspace, []string{"-80", "80-"}, 0, false); err != nil {
+			return 1
+		}
+
+		// Start vtgate
+		if err := clusterInstance.StartVtgate(); err != nil {
+			return 1
+		}
+
+		vtParams = mysql.ConnParams{
+			Host: "localhost",
+			Port: clusterInstance.VtgateMySQLPort,
+		}
+
+		conn, closer, err := utils.NewMySQL(clusterInstance, "mks", unshardedSQLSchema+shardedSQLSchema)
+		if err != nil {
+			return 1
+		}
+		defer closer()
+		mysqlParams = conn
+
+		// Materialize zip_detail to sharded keyspace.
+		if err := clusterInstance.VtctlProcess.ExecuteCommand(
+			"Materialize",
+			"--",
+			"--tablet_types",
+			"PRIMARY",
+			`{
+				"workflow": "materialize_zip_detail",
+				"source_keyspace": "`+unshardedKeyspaceName+`",
+				"target_keyspace": "`+shardedKeyspaceName+`",
+				"table_settings": [
+					{
+						"target_table": "zip_detail",
+						"source_expression": "select * from zip_detail"
+					}
+				],
+				"stop_after_copy": true
+			}`,
+		); err != nil {
+			return 1
+		}
+
+		ctx := context.Background()
+		vtgateAddr := fmt.Sprintf("%s:%d", clusterInstance.Hostname, clusterInstance.VtgateProcess.GrpcPort)
+		vtgateConn, err := vtgateconn.Dial(ctx, vtgateAddr)
+		if err != nil {
+			return 1
+		}
+
+		// INSERT some delivery_failure rows.
+		session := vtgateConn.Session("@primary", nil)
+		if _, err := session.Execute(ctx, `
+			INSERT INTO delivery_failure(id, zip_detail_id, reason)
+			VALUES (1, 1, 'Failed delivery due to discontinued zipcode.'),
+			       (2, 2, 'Failed delivery due to discontinued zipcode.'),
+			       (3, 3, 'Failed delivery due to unknown reason.');
+		`, map[string]*querypb.BindVariable{}); err != nil {
+			return 1
+		}
+
+		// TODO(maxeng) remove when we have a proper way to check
+		// materialization lag and cutover.
```

Comment:
Just FYI, if it's helpful:
  - https://github.com/vitessio/vitess/blob/main/go/test/endtoend/vreplication/helper_test.go#L154
  - https://github.com/vitessio/vitess/blob/main/go/test/endtoend/vreplication/helper_test.go#L245
  - https://github.com/vitessio/vitess/blob/main/go/test/endtoend/vreplication/helper_test.go#L210